Userland/Libraries/LibWeb/Layout/GridFormattingContext.cpp

```diff
@@ -1582,62 +1570,25 @@ int GridFormattingContext::get_line_index_by_line_name(String const& needle, CSS
     return -1;
 }
 
-OccupationGrid::OccupationGrid(size_t column_count, size_t row_count)
-{
-    Vector<bool> occupation_grid_row;
-    for (size_t column_index = 0; column_index < max(column_count, 1); column_index++)
-        occupation_grid_row.append(false);
-    for (size_t row_index = 0; row_index < max(row_count, 1); row_index++)
-        m_occupation_grid.append(occupation_grid_row);
-}
-
-OccupationGrid::OccupationGrid()
-{
-}
-
-void OccupationGrid::maybe_add_column(size_t needed_number_of_columns)
-{
-    if (needed_number_of_columns <= column_count())
-        return;
-    auto column_count_before_modification = column_count();
-    for (auto& occupation_grid_row : m_occupation_grid)
-        for (size_t idx = 0; idx < needed_number_of_columns - column_count_before_modification; idx++)
-            occupation_grid_row.append(false);
-}
-
-void OccupationGrid::maybe_add_row(size_t needed_number_of_rows)
-{
-    if (needed_number_of_rows <= row_count())
-        return;
-
-    Vector<bool> new_occupation_grid_row;
-    for (size_t idx = 0; idx < column_count(); idx++)
-        new_occupation_grid_row.append(false);
-
-    for (size_t idx = 0; idx < needed_number_of_rows - row_count(); idx++)
-        m_occupation_grid.append(new_occupation_grid_row);
-}
-
 void OccupationGrid::set_occupied(size_t column_start, size_t column_end, size_t row_start, size_t row_end)
 {
-    for (size_t row_index = 0; row_index < row_count(); row_index++) {
-        if (row_index >= row_start && row_index < row_end) {
-            for (size_t column_index = 0; column_index < column_count(); column_index++) {
-                if (column_index >= column_start && column_index < column_end)
-                    set_occupied(column_index, row_index);
-            }
+    for (size_t row = row_start; row < row_end; row++) {
+        for (size_t column = column_start; column < column_end; column++) {
+            set_occupied(column, row);
         }
     }
 }
 
 void OccupationGrid::set_occupied(size_t column_index, size_t row_index)
 {
-    m_occupation_grid[row_index][column_index] = true;
+    m_columns_count = max(m_columns_count, column_index + 1);
+    m_rows_count = max(m_rows_count, row_index + 1);
+    m_occupation_grid.set(GridPosition { .row = row_index, .column = column_index }, true);
```

Comment:
1. If you're only setting values to `true`, consider using `HashTable`
2. Please use the fallible versions if they exist, i.e. `try_set`. This will expose allocation issues, for example.

Reply:
> Please use the fallible versions if they exist, i.e. try_set. This will expose allocation issues, for example.

using fallible versions makes perfect sense but propagating errors from `FormattingContext::run()` is a bit beyond the scope of this change :)

Reply:
Nobody asked to propagate errors from `FormattingContext::run()`? :^)

We have `.release_value_but_fixme_should_propagate_errors()` so you have an explicit FIXME invocation in there.